### PR TITLE
[FLINK-2655] Minimize intermediate merging of spilled buffers

### PR DESCRIPTION
**Pull request by greghogan**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/UnilateralSortMerger.java [hunk 1]
```diff
@@ -1525,34 +1525,41 @@ protected final void disposeSortBuffers(boolean releaseMemory) {
...
-			
+			// A channel list with length maxFanIn<sup>i</sup> can be merged to maxFanIn files in i-1 rounds where every merge
+			// is a full merge with maxFanIn input channels. A partial round includes merges with fewer than maxFanIn
+			// inputs. It is most efficient to perform the partial round first.
+			final double scale = Math.ceil(Math.log(channelIDs.size()) / Math.log(this.maxFanIn)) - 1;
+
+			final int numStart = channelIDs.size();
+			final int numEnd = (int) Math.pow(this.maxFanIn, scale);
+
+			final int numMerges = (int) Math.ceil((numStart - numEnd) / (double) (this.maxFanIn - 1));
```
Comment (HuangWHWHW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/UnilateralSortMerger.java]: Are you sure the `this.maxFanIn` will never equal 1(avoid dividing 0)?

Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/UnilateralSortMerger.java]: Yes, `maxFanIn` is a configuration parameter and must be > 1.

Comment (HuangWHWHW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/UnilateralSortMerger.java]: Ah, thanks for this info :)

**Comment by fhueske** [comment]
Thanks for the PR @greghogan.
Good to merge, IMO.

**Comment by fhueske** [comment]
Will merge this PR tomorrow.

**Pull request closed without merging**